gtsam/hybrid/tests/testGaussianMixtureFactor.cpp

```diff
@@ -387,50 +389,91 @@ TEST(GaussianMixtureFactor, GaussianMixtureModel2) {
-static HybridBayesNet CreateBayesNet(double mu0, double mu1, double sigma0,
-                                     double sigma1,
-                                     bool add_second_measurement = false,
-                                     double prior_sigma = 1e-3,
-                                     double measurement_sigma = 3.0) {
-  DiscreteKey m1(M(1), 2);
-  Key z0 = Z(0), z1 = Z(1);
-  Key x0 = X(0), x1 = X(1);
-
-  HybridBayesNet hbn;
-
-  auto measurement_model = noiseModel::Isotropic::Sigma(1, measurement_sigma);
-  // Add measurement P(z0 | x0)
-  auto p_z0 = std::make_shared<GaussianConditional>(
-      z0, Vector1(0.0), -I_1x1, x0, I_1x1, measurement_model);
-  hbn.push_back(p_z0);
+DiscreteKey m1(M(1), 2);
 
-  // Add hybrid motion model
+/// Create hybrid motion model p(x1 | x0, m1)
+static GaussianMixture::shared_ptr CreateHybridMotionModel(double mu0,
+                                                           double mu1,
+                                                           double sigma0,
+                                                           double sigma1) {
   auto model0 = noiseModel::Isotropic::Sigma(1, sigma0);
   auto model1 = noiseModel::Isotropic::Sigma(1, sigma1);
-  auto c0 = make_shared<GaussianConditional>(x1, Vector1(mu0), I_1x1, x0,
+  auto c0 = make_shared<GaussianConditional>(X(1), Vector1(mu0), I_1x1, X(0),
                                              -I_1x1, model0),
-       c1 = make_shared<GaussianConditional>(x1, Vector1(mu1), I_1x1, x0,
+       c1 = make_shared<GaussianConditional>(X(1), Vector1(mu1), I_1x1, X(0),
                                              -I_1x1, model1);
+  return std::make_shared<GaussianMixture>(
+      KeyVector{X(1)}, KeyVector{X(0)}, DiscreteKeys{m1}, std::vector{c0, c1});
+}
+
+/// Create two state Bayes network with 1 or two measurement models
+HybridBayesNet CreateBayesNet(
+    const GaussianMixture::shared_ptr& hybridMotionModel,
+    bool add_second_measurement = false) {
+  HybridBayesNet hbn;
 
-  auto motion = std::make_shared<GaussianMixture>(
-      KeyVector{x1}, KeyVector{x0}, DiscreteKeys{m1}, std::vector{c0, c1});
-  hbn.push_back(motion);
+  // Add measurement model p(z0 | x0)
+  const double measurement_sigma = 3.0;
+  auto measurement_model = noiseModel::Isotropic::Sigma(1, measurement_sigma);
+  hbn.emplace_shared<GaussianConditional>(Z(0), Vector1(0.0), I_1x1, X(0),
+                                          -I_1x1, measurement_model);
 
+  // Optionally add second measurement model p(z1 | x1)
   if (add_second_measurement) {
-    // Add second measurement
-    auto p_z1 = std::make_shared<GaussianConditional>(
-        z1, Vector1(0.0), -I_1x1, x1, I_1x1, measurement_model);
-    hbn.push_back(p_z1);
+    hbn.emplace_shared<GaussianConditional>(Z(1), Vector1(0.0), I_1x1, X(1),
+                                            -I_1x1, measurement_model);
   }
 
+  // Add hybrid motion model
+  hbn.push_back(hybridMotionModel);
```

Comment:
Also, motion model needs to be added here otherwise sample (which depends on topological sort order) fails.